src/common/src/array/vis.rs

```diff
@@ -0,0 +1,259 @@
+    pub fn iter(self) -> impl Iterator<Item = bool> + 'a {
+        match self {
+            VisRef::Bitmap(b) => b.iter(),
+            VisRef::Compact(c) => iter::repeat(true).take(c),
+        }
+    }
+
+    #[auto_enum(Iterator)]
+    pub fn ones(self) -> impl Iterator<Item = usize> + 'a {
+        match self {
+            VisRef::Bitmap(b) => b.ones(),
+            VisRef::Compact(c) => 0..c,
+        }
+    }
+}
+
+impl<'a> From<&'a Bitmap> for VisRef<'a> {
+    fn from(b: &'a Bitmap) -> Self {
+        VisRef::Bitmap(b)
+    }
+}
+
+impl<'a> From<usize> for VisRef<'a> {
+    fn from(c: usize) -> Self {
+        VisRef::Compact(c)
+    }
+}
+
+impl<'a> From<&'a Vis> for VisRef<'a> {
+    fn from(vis: &'a Vis) -> Self {
+        vis.as_ref()
+    }
+}
+
+impl<'a, 'b> std::ops::BitAnd<VisRef<'b>> for VisRef<'a> {
+    type Output = Vis;
+
+    fn bitand(self, rhs: VisRef<'b>) -> Self::Output {
+        match (self, rhs) {
+            (VisRef::Bitmap(b1), VisRef::Bitmap(b2)) => Vis::Bitmap(b1 & b2),
+            (VisRef::Bitmap(b1), VisRef::Compact(c2)) => {
+                assert_eq!(b1.len(), c2);
+                Vis::Bitmap(b1.clone())
+            }
+            (VisRef::Compact(c1), VisRef::Bitmap(b2)) => {
+                assert_eq!(c1, b2.len());
+                Vis::Bitmap(b2.clone())
+            }
+            (VisRef::Compact(c1), VisRef::Compact(c2)) => {
+                assert_eq!(c1, c2);
+                Vis::Compact(c1)
+            }
+        }
+    }
+}
+
+impl<'a> std::ops::BitAnd<Vis> for VisRef<'a> {
+    type Output = Vis;
+
+    fn bitand(self, rhs: Vis) -> Self::Output {
+        match (self, rhs) {
+            (VisRef::Bitmap(b1), Vis::Bitmap(b2)) => Vis::Bitmap(b1 & b2),
+            (VisRef::Bitmap(b1), Vis::Compact(c2)) => {
+                assert_eq!(b1.len(), c2);
+                Vis::Bitmap(b1.clone())
+            }
+            (VisRef::Compact(c1), Vis::Bitmap(b2)) => {
+                assert_eq!(c1, b2.len());
+                Vis::Bitmap(b2)
+            }
+            (VisRef::Compact(c1), Vis::Compact(c2)) => {
+                assert_eq!(c1, c2);
+                Vis::Compact(c1)
+            }
+        }
+    }
+}
+
+impl<'a, 'b> std::ops::BitOr<VisRef<'b>> for VisRef<'a> {
+    type Output = Vis;
+
+    fn bitor(self, rhs: VisRef<'b>) -> Self::Output {
+        match (self, rhs) {
+            (VisRef::Bitmap(b1), VisRef::Bitmap(b2)) => Vis::Bitmap(b1 | b2),
+            (VisRef::Bitmap(b1), VisRef::Compact(c2)) => {
+                assert_eq!(b1.len(), c2);
+                Vis::Compact(c2)
+            }
+            (VisRef::Compact(c1), VisRef::Bitmap(b2)) => {
+                assert_eq!(c1, b2.len());
+                Vis::Compact(c1)
+            }
+            (VisRef::Compact(c1), VisRef::Compact(c2)) => {
+                assert_eq!(c1, c2);
+                Vis::Compact(c1)
+            }
+        }
+    }
+}
+
+impl<'a> std::ops::BitOr<Vis> for VisRef<'a> {
```

Comment:
Didn't fully understant the variants yet 😇. Could `impl<'a> std::ops::BitOr<Vis> for Vis` also saves a clone?

Reply:
For Bitand, yes, but we don’t use it now, so we haven’t implemented it.

For Bitor, no.